gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+          ResourceRecordSet wrappedRrset = wrapper.rrset();
+          if (rrset.getName().equals(wrappedRrset.getName())
+              && rrset.getType().equals(wrappedRrset.getType())) {
+            // such a record exist and we must have a deletion
+            if (deletions == null || !deletions.contains(wrappedRrset)) {
+              return Error.ALREADY_EXISTS.response(String.format(
+                  "The 'entity.change.additions[%s]' resource named '%s (%s)' already exists.",
+                  index, rrset.getName(), rrset.getType()));
+            }
+          }
+        }
+        if (rrset.getType().equals("SOA") && findByNameAndType(deletions, null, "SOA") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.additions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'SOA' at the apex.", index));
+        }
+        if (rrset.getType().equals("NS") && findByNameAndType(deletions, null, "NS") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.additions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'NS' at the apex.", index));
+        }
+        index++;
+      }
+    }
+    if (deletions != null) {
+      int index = 0;
+      for (ResourceRecordSet rrset : deletions) {
+        if (rrset.getType().equals("SOA") && findByNameAndType(additions, null, "SOA") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.deletions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'SOA' at the apex.", index));
+        }
+        if (rrset.getType().equals("NS") && findByNameAndType(additions, null, "NS") == null) {
+          return Error.INVALID_ZONE_APEX.response(String.format("The resource record set 'entity"
+              + ".change.deletions[%s]' is invalid because a zone must contain exactly one resource"
+              + " record set of type 'NS' at the apex.", index));
+        }
+        index++;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Helper for searching rrsets in a collection.
+   */
+  private static ResourceRecordSet findByNameAndType(Iterable<ResourceRecordSet> records,
+      String name, String type) {
+    if (records != null) {
+      for (ResourceRecordSet rrset : records) {
+        if ((name == null || name.equals(rrset.getName()))
+            && (type == null || type.equals(rrset.getType()))) {
+          return rrset;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * We only provide the most basic validation for A and AAAA records.
+   */
+  static boolean checkRrData(String data, String type) {
+    // todo add validation for other records
```

Comment:
Just my opinion, but it seems like we may not want to be in the business of validating rrdata beyond simple checks (e.g., is a field null when it isn't supposed to be).  It seems like we could spend a lot of time here trying to meticulously mapping out what data the service accepts, but I don't think there's much gain in doing that.